documents/design/oavb_ipc.asciidoc

```diff
@@ -0,0 +1,83 @@
+= Interprocess Communication Recommendations for AVB/TSN
+Andrew Elder
+v0.1
+
+== Introduction
+
+The AVB/TSN software stack is comprised of a number of modules that implement protocol stacks specified by IEEE standards. Each of these protocols implement an Ethernet network facing interface and an application facing interface. The required application facing interface is described in some detail in http://avnu.org/wp-content/uploads/2014/05/AVnu_SWAPIs_v1.0.pdf[AVB Software Interfaces and Endpoint Architecture Guidelines]. Since a module running a particular protocol may be a standalone execution unit (for example a Linux daemon), an interprocess communication method is required to interface to the execution unit.
+
+For portability reasons it is recommended to separate the implementation of the interprocess communication method as much as possible from the logic of the protocol implementation. This applies to both an application that is interfacing to an AVB/TSN module and the implementation of the AVB/TSN module itself.
+
+For the purposes of this document the following straw-man application interfaces are described:
+
+.Straw-man Interfaces
+[source,c/c++]
+----
+
+void AVBTSN_SetParam(uint32_t param);
+
+----
+
+where the AVBTSN portion of the function name would be replaced by the name of the IEEE AVB/TSN protocol module that is being targeted.
```

Comment:
One difficulty that can arise with get/set-based interfaces is the situation when some set of changes need to be made more-or-less simultaneously across a set of parameters. You typically need some way to encapsulate a complete transaction or provide the same sort of functionality at a lower level, such as a 'shadow' block of parameters along with a 'trigger' that commits the shadow block to the actual configuration.

Do you intend for this interface to use a uniform datatype like uint32_t for parameters, or will the parameter type vary by parameter? What about communicating failure to get/set a parameter? Will responses to commands be communicated synchronously or asynchronously, or will it be up to the implementation?

Some of the protocol modules can expose quite a few parameters, and in some cases duplicates of the same set of parameters over multiple instances of some interface. E.g. ports or streams. It might make sense to impose a hierarchy on the interface identifiers, at least at the abstract interface description level.